PovioKit/Classes/ShareKit/Asset Library/SKAssetLibraryProvider.swift

```diff
@@ -0,0 +1,152 @@
+  }
+  private var assetCache = [AssetType: PHAsset]()
+  
+  func authorizeIfNecessary(success: (() -> Void)?, failure: (() -> Void)?) {
+    guard PHPhotoLibrary.authorizationStatus() != .authorized else {
+      success?()
+      return
+    }
+    
+    PHPhotoLibrary.requestAuthorization { status in
+      status == .authorized ? success?() : failure?()
+    }
+  }
+  
+  func findOrCreateAlbum(name: String, completion: @escaping (PHAssetCollection?) -> Void) {
+    let fetchOptions = PHFetchOptions()
+    fetchOptions.predicate = NSPredicate(format: "title = %@", name)
+    guard let photoAlbum = PHAssetCollection.fetchAssetCollections(with: .album, subtype: .any, options: fetchOptions).firstObject else {
+      return createAlbum(name: name, completion: completion)
+    }
+    return completion(photoAlbum)
+  }
+  
+  func createAlbum(name: String, completion: @escaping (PHAssetCollection?) -> Void) {
+    var albumPlaceholder: PHObjectPlaceholder?
+    PHPhotoLibrary.shared().performChanges({
+      let createAlbumRequest = PHAssetCollectionChangeRequest.creationRequestForAssetCollection(withTitle: name)
+      albumPlaceholder = createAlbumRequest.placeholderForCreatedAssetCollection
+    }, completionHandler: { success, error in
+      guard success else {
+        return completion(nil)
+      }
+      guard let placeholder = albumPlaceholder else {
+        completion(nil)
+        return
+      }
+      guard let album = PHAssetCollection.fetchAssetCollections(withLocalIdentifiers: [placeholder.localIdentifier], options: nil).firstObject else {
+        completion(nil)
+        return
+      }
+      completion(album)
+    })
+  }
+  
+  func saveImage(image: UIImage, album: PHAssetCollection, completion: @escaping (PHAsset?) -> Void) {
+    if let asset = assetCache[AssetType.image(image)] {
+      completion(asset)
+      return
+    }
+    
+    var placeholder: PHObjectPlaceholder?
+    PHPhotoLibrary.shared().performChanges({
+      let createAssetRequest = PHAssetChangeRequest.creationRequestForAsset(from: image)
+      guard let albumChangeRequest = PHAssetCollectionChangeRequest(for: album),
+        let photoPlaceholder = createAssetRequest.placeholderForCreatedAsset else { return }
+      placeholder = photoPlaceholder
+      let fastEnumeration = NSArray(array: [photoPlaceholder] as [PHObjectPlaceholder])
+      albumChangeRequest.addAssets(fastEnumeration)
+    }, completionHandler: { success, error in
+      guard let placeholder = placeholder else {
+        completion(nil)
+        return
+      }
+      if success {
+        let assets = PHAsset.fetchAssets(withLocalIdentifiers: [placeholder.localIdentifier], options: nil)
+        self.assetCache[AssetType.image(image)] = assets.firstObject
+        completion(assets.firstObject)
+      } else {
+        completion(nil)
+      }
+    })
+  }
+  
+  func saveVideo(filePath path: String, album: PHAssetCollection, completion: @escaping (PHAsset?) -> Void) {
```

Comment:
These 2 methods differ very slightly, any chance of making one generic one that takes the asset type & creation request? (since those 2 are the only lines that differ, by first glance)